src/repository.c

```diff
@@ -790,48 +843,87 @@ static int repo_is_worktree(unsigned *out, const git_repository *repo)
 	return error;
 }
 
-int git_repository_open_ext(
+int git_repository_open_init_options(
+	git_repository_open_options *opts,
+	unsigned int version)
+{
+	GIT_INIT_STRUCTURE_FROM_TEMPLATE(
+		opts, version, git_repository_open_options,
+		GIT_REPOSITORY_OPEN_OPTIONS_INIT);
+	return 0;
+}
+
+int git_repository_open_with_opts(
 	git_repository **repo_ptr,
 	const char *start_path,
-	unsigned int flags,
-	const char *ceiling_dirs)
+	const git_repository_open_options *given_opts)
 {
 	int error;
 	unsigned is_worktree;
-	git_buf gitdir = GIT_BUF_INIT, workdir = GIT_BUF_INIT,
-		gitlink = GIT_BUF_INIT, commondir = GIT_BUF_INIT;
 	git_repository *repo = NULL;
 	git_config *config = NULL;
+	git_repository_open_options opts = GIT_REPOSITORY_OPEN_OPTIONS_INIT;
+	struct repo_env env;
+
+	GIT_ERROR_CHECK_VERSION(
+		given_opts, GIT_REPOSITORY_OPEN_OPTIONS_VERSION, "git_repository_open_options");
+
+	if (given_opts)
+		memcpy(&opts, given_opts, sizeof(opts));
+
+//	if ((error = repo_env_init(&env, start_path, &opts)) < 0 ||
+//		(error = repo_env_find(&env, start_path)) < 0 ||
+//		!repo_ptr)
+//		goto cleanup;
+
+	if ((error = repo_env_init(&env, start_path, &opts)) < 0)
+		goto cleanup;
 
-	if (flags & GIT_REPOSITORY_OPEN_FROM_ENV)
-		return _git_repository_open_ext_from_env(repo_ptr, start_path);
+	fprintf(stderr, "init\n");
+	repo_env_dump(&env);
 
-	if (repo_ptr)
-		*repo_ptr = NULL;
+	if ((error = repo_env_find(&env, start_path)) < 0)
+		goto cleanup;
 
-	error = find_repo(
-		&gitdir, &workdir, &gitlink, &commondir, start_path, flags, ceiling_dirs);
+	fprintf(stderr, "find\n");
+	repo_env_dump(&env);
 
-	if (error < 0 || !repo_ptr)
+	if (!repo_ptr)
 		goto cleanup;
 
+	/* The caller really want to open the repository, proceed */
+
+	*repo_ptr = NULL;
+
 	repo = repository_alloc();
 	GIT_ERROR_CHECK_ALLOC(repo);
 
-	repo->gitdir = git_buf_detach(&gitdir);
+	repo->gitdir = git_buf_detach(&env.gitdir_path);
 	GIT_ERROR_CHECK_ALLOC(repo->gitdir);
 
-	if (gitlink.size) {
-		repo->gitlink = git_buf_detach(&gitlink);
+	if (git_buf_len(&env.gitlink_path) > 0) {
+		repo->gitlink = git_buf_detach(&env.gitlink_path);
 		GIT_ERROR_CHECK_ALLOC(repo->gitlink);
 	}
-	if (commondir.size) {
-		repo->commondir = git_buf_detach(&commondir);
+
+	if (git_buf_len(&env.commondir_path) > 0) {
+		repo->commondir = git_buf_detach(&env.commondir_path);
 		GIT_ERROR_CHECK_ALLOC(repo->commondir);
 	}
 
+//	if (git_buf_len(&env.workdir_path) > 0) {
+//		repo->workdir = git_buf_detach(&env.workdir_path);
+//		GIT_ERROR_CHECK_ALLOC(repo->workdir)
+//	}
+
+	if (git_buf_len(&env.configfile_path) > 0) {
+		repo->configfile = git_buf_detach(&env.configfile_path);
+		GIT_ERROR_CHECK_ALLOC(repo->configfile);
+	}
+
 	if ((error = repo_is_worktree(&is_worktree, repo)) < 0)
```

Comment:
[Function](https://github.com/libgit2/libgit2/blob/master/src/repository.c#L770).

This could have easily been part of `repo_env_find`, since it only look at the filesystem.